staging/src/k8s.io/code-generator/cmd/lister-gen/generators/lister.go

```diff
@@ -345,35 +321,13 @@ type $.type|public$NamespaceLister interface {
 }
 `
 
+// This embeds a typed namespaced resource indexer instead of aliasing, so that the struct
+// is available as a receiver for methods specific to the generated type
+// (from the corresponding expansion interface).
 var namespaceListerStruct = `
 // $.type|private$NamespaceLister implements the $.type|public$NamespaceLister
 // interface.
 type $.type|private$NamespaceLister struct {
```

Comment:
The code is identical but the purposes are different. The private structs are hooks for expansions, and if we merge the two then the functions from both leak into each other (in particular, the function in the non-namespaced lister that gives a namespaced lister)...

Of course I can’t find any actual usage of expansions on listers so perhaps we could just delete that mechanism wholesale.

Reply:
Not sure I follow.  Let's pick on Service as an example.

if I add a `FooBar` method to `ServiceListerExpansion`, then `NewServiceLister()` (which returns a generated wrapper of `listers.ResourceIndexer[*v1.Service]`) fails to compile because  `ResourceIndexer[*v1.Service]` does not have a `FooBar` method and no longer satisfies `ServiceLister`.

How would someone use the extensions?  I'm not seeing it.

Reply:
It works in a similar fashion to `client-gen` expansions; see `pod_expansion.go` for example.

For a Service lister, you’d add `service_expansion.go` containing

```go
package v1

// ServiceListerExpansion allows custom methods to be added to
// ServiceLister.
type ServiceListerExpansion interface {
	FooBar()
}

// ServiceNamespaceListerExpansion allows custom methods to be added to
// ServiceNamespaceLister.
type ServiceNamespaceListerExpansion interface {
}

func (s *serviceLister) FooBar() {
	// Play that funky music
}
```

Then run `lister-gen`, which will notice that there is an expansion defined for the Service lister, and skip it when generating `expansion_generated.go`. Then everything compiles, and `NewServiceLister()` returns an instance with a `FooBar` implementation alongside the other methods.

(The two expansion interfaces are handled together by `lister-gen`; if one is defined manually, the other needs to be too.)

Reply:
I’ve added an expansion example in the codegen examples.

Reply:
You’re welcome! I’d rather spend time discussing these things than have PRs approved without detailed review… Next up is #121439 ;-)